power_grid_model_c/power_grid_model/include/power_grid_model/component/base.hpp

```diff
@@ -34,6 +34,11 @@ class Base {
     }
     virtual bool energized(bool is_connected_to_source) const = 0;
 
+    Base(const Base&) = default;
+    Base& operator=(const Base&) = default;
+    Base(Base&&) = default;
+    Base& operator=(Base&&) = default;
+
```

Comment:
instead of adding all `default`, the default `virtual deleter` should be removed in line `28` (`virtual ~Base() = default;`).

Remember: the rule of three-five-zero prefers having zero specializations over five. only if the rule of zero cannot be used should the rule of five be followed

```suggestion
```

Reply:
If i remove all five of them and rebuild validation tests, I get 18 more warnings treated as errors. I suppose leaving it as it is now might be a good idea?

Reply:
Oh wait my bad, since you explicitly mark it `virtual` and `public`, it **is** needed (cfr. explanation of rule of zero in https://en.cppreference.com/w/cpp/language/rule_of_three )

see also https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c67-a-polymorphic-class-should-suppress-public-copymove

let me investigate how to work around this

Reply:
ok i understand the reason now. you **do** need to have all declared, but please make the copy and copy assignment constructors `protected`. You will likely also have to explicitly declare the copy and copy assignment constructors in the individual components that are **NOT** intended for inheritance (the ones marked `final`, e.g. Line rather than in Branch) as `public`.

Reply:
please note that this is one of the many reasons why people are moving away from polymorphism models (inheritance) 😬